### PR TITLE
feat(detectors): add ClickHouse detector

### LINKED ISSUE #5337: Add a ClickHouse detector
### Proposal

Add a detector for ClickHouse credentials. ClickHouse hosts data and has a paid cloud tier, so it looks like a fit for the sourcing guidelines in `hack/docs/Adding_Detectors_external.md`.

There is no ClickHouse detector today, and I couldn't find a prior issue or PR proposing one. `pkg/sources/elasticsearch` shows the project already treats analytical stores as worth covering as scan targets; this is the credential side for ClickHouse.

### What it would match

Two forms, both of which carry the host alongside the credential, so they can be verified the way the existing `postgres`, `mongodb` and `redis` detectors are:

- Native connection strings — `clickhouse://user:pass@host:9000/db`, and `clickhouses://` for TLS. This is the form the Go, Python and JDBC clients all accept.
- ClickHouse Cloud, which speaks HTTPS rather than the native protocol, so its credentials appear as ordinary basic-auth URLs. Pinning these to the `clickhouse.cloud` domain keeps it high-signal — a bare `https://user:pass@host` URL is not a ClickHouse finding.

### Verification

Over ClickHouse's HTTP interface with `SELECT 1`, authenticating with the `X-ClickHouse-User` and `X-ClickHouse-Key` headers. That keeps it to `net/http` with no database driver added, and works for both self-hosted and Cloud.

I checked the response codes against ClickHouse 26.8: `200` for a working credential, and `403` for both a wrong password and an unknown user. So those are determinate and anything else returns an error.

### The one design question

Connection strings normally carry a native port (9000, or 9440 for TLS) while the HTTP interface listens on 8123/8443, so verification has to map between them. I've mapped the known native ports onto their HTTP equivalents and left anything else alone, since some deployments publish the HTTP interface on a different port. Where the HTTP interface isn't reachable the request simply fails, which surfaces as indeterminate rather than as a claim that the credential is invalid.

The alternative is speaking the native protocol, which would mean taking on a ClickHouse driver dependency. That seemed worse than a port mapping given the guidance to favour `net/http`, but happy to switch if you'd rather.

Happy to take feedback on any of the above. A PR is ready.

### PR DESCRIPTION
**Pull request by DrFaust92**
Closes #5337.

## What

Adds a detector for ClickHouse credentials, covering two forms:

- **Native connection strings** — `clickhouse://user:pass@host:9000/db` and `clickhouses://` for TLS, the form accepted by the Go, Python and JDBC clients.
- **ClickHouse Cloud** — which speaks HTTPS rather than the native protocol, so credentials appear as basic-auth URLs. These are pinned to the `clickhouse.cloud` domain so that an ordinary `https://user:pass@host` URL is not reported.

`SecretParts` carries `host`, `username` and `password`; `ExtraData` carries `host`, `username` and the database when the connection string names one. Results are deduplicated per distinct credential rather than per occurrence, and passwords that are all `*` (redacted) or start with `$` (an unexpanded variable) are skipped.

## Verification

Over ClickHouse's HTTP interface with `SELECT 1`, authenticating with the `X-ClickHouse-User` and `X-ClickHouse-Key` headers. This keeps the detector on `net/http` with no database driver added, and works for both self-hosted servers and Cloud.

I checked the behaviour against ClickHouse 26.8 rather than assuming it: `200` for a working credential, `403` for both a wrong password and an unknown user. Those are treated as determinate; anything else returns an error so the failure is indeterminate.

Connection strings normally carry a native port, so 9000 and 9440 are mapped onto the HTTP interface's 8123 and 8443. A port that is neither is left alone, since some deployments publish the HTTP interface elsewhere — and where it isn't reachable at all the request just fails, which is indeterminate rather than a claim that the credential is invalid. This is the part I'd most welcome an opinion on; the alternative is taking on a ClickHouse driver to speak the native protocol, which seemed worse than a port mapping.

## Testing

`go test ./pkg/detectors/clickhouse/...` — 24 subtests across three groups:

- pattern tests for both forms, including a connection string with no username (ClickHouse's `default`), duplicate suppression, and negative cases for redacted passwords, unexpanded variables, short passwords, non-Cloud HTTPS URLs and URLs with no credentials
- the native-to-HTTP port mapping
- verification against an `httptest` server for 200/401/403/500/502, asserting the auth headers and query that get sent, so it needs no container in CI

`go test ./pkg/engine/defaults/...` passes, and `gofmt`/`go vet` are clean. I also confirmed the verification path end to end against a real ClickHouse server in Docker, including that a connection string on port 9000 verifies through the mapped HTTP port.

---

> [!NOTE]
> **Medium Risk**
> Verification performs outbound HTTP to hosts taken from leaked connection strings and transmits extracted passwords, though Cloud hostname and ClickHouse response-header checks reduce sending creds to unrelated servers.
> 
> **Overview**
> Adds a **ClickHouse** secret detector and wires it into the default engine and `DetectorType` enum (`ClickHouse = 1072`).
> 
> It finds credentials in **`clickhouse://` / `clickhouses://`** connection strings and in **`https://user:pass@…clickhouse.cloud`** URLs (HTTPS matches are limited to true Cloud hostnames, not generic basic-auth links). Parsing decodes URL-encoded userinfo, defaults missing usernames to `default`, skips redacted `*` passwords and `$…` placeholders, dedupes by host/user/password, and exposes host, username, password (and database when present) on results.
> 
> When verification is enabled, it maps common native ports to ClickHouse’s HTTP interface and runs **`SELECT 1`** using **`X-ClickHouse-User`** / **`X-ClickHouse-Key`**, treating **200** as verified and **401/403** as failed only when the response looks like ClickHouse (`X-ClickHouse-*` headers). Tests cover matching edge cases, port mapping, and verify behavior via `httptest`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 12b4396dd8ef1e646a1c00807d799b3ebd9d9f3c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by DrFaust92** [comment]
Thanks — all three were real, and I checked each against a live ClickHouse 26.8 rather than just reading the code. Fixed in 12b4396.

**Encoded credentials fail verification.** Confirmed, and the worst of the three. With a user whose password is `p@ssw0rd!x`, the server answers `200` for the raw value and `403` for `p%40ssw0rd%21x` — and since the pattern's character class excludes a literal `@`, the escaped form is the only way such a password can appear in a connection string. So a correctly written, working credential was being reported as determinately invalid. Userinfo is now percent-decoded before it is used for verification, `SecretParts` or `Raw`; `Redacted` still masks the password as it was written. A value that isn't validly encoded is left alone, so a password containing a bare `%` still reports as written.

**Any HTTP 200 verifies credentials.** Also correct — the host comes from the match and an unrecognised port is passed through untouched, so the responder need not be ClickHouse. Verification now requires an `X-ClickHouse-*` header before reading anything into the status. Real replies carry them in both directions, which I verified: a 200 returns `X-ClickHouse-Query-Id` and `X-ClickHouse-Summary`, and a 403 returns `X-ClickHouse-Exception-Code: 516`. Absent any of them the result is indeterminate. That also covers the converse case, where a 403 from an unrelated server would previously have read as a rejected credential.

**Cloud host match is unanchored.** Correct. `foo.clickhouse.cloud.example.com` matched and was truncated to `foo.clickhouse.cloud`, so verification would have sent the credential to a Cloud hostname it does not belong to — worth a bit more than Low for that reason. go-re2 has no lookahead, so rather than contort the pattern the host is now captured whole and the domain checked with a suffix test in code.

Tests are up from 24 to 29 subtests: percent-decoded password and user, a Cloud lookalike host, and non-ClickHouse 200/403 responses. `gofmt` and `go vet` are clean, and I re-ran the live check — the encoded password now verifies, and a wrong password still comes back unverified with no error.

File: pkg/detectors/clickhouse/clickhouse.go [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+// isCloudHost reports whether a host belongs to ClickHouse Cloud, matching on the
+// domain rather than a substring so that a lookalike host is not accepted.
+func isCloudHost(hostPort string) bool {
+	host := hostPort
+	if h, _, err := net.SplitHostPort(hostPort); err == nil {
+		host = h
+	}
+	return strings.HasSuffix(strings.ToLower(host), cloudDomain)
```
[the remark this answers is not in the document]
Comment (DrFaust92) [reply pkg/detectors/clickhouse/clickhouse.go:182]: Good catch, and it was a regression I introduced in 12b4396 — moving the domain check out of the pattern is exactly what made the trailing dot start mattering, since the host group then swallowed it and the suffix test rejected the result.

Confirmed before fixing. A portless Cloud URL ending a sentence returned zero results, where previously the stray dot was simply left unmatched. The native form still reported, but carried the dot into the host it then verified against.

Fixed in af17c66 by trimming trailing dots from the captured host, handling the `host.:9000` form via `SplitHostPort` rather than trimming the whole string. That is also the right answer for a genuinely absolute name, which resolves the same either way.

Tests are up from 29 to 37 subtests: both prose cases, plus a `trimHostDots` table covering ports, repeated dots and a dots-only host.

File: pkg/detectors/clickhouse/clickhouse.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	// Connection strings using ClickHouse's own scheme, which is what the Go,
+	// Python and JDBC clients all accept: clickhouse://user:pass@host:9000/db
+	// (clickhouses:// for TLS). Capture groups: user, password, host[:port], /database.
+	keyPat = regexp.MustCompile(`\bclickhouses?://([^\s:/@]{0,64}):([^\s:/@]{3,100})@([-.\w]+(?::\d{1,5})?)(/[^\s?#"'` + "`" + `]*)?`)
+
+	// ClickHouse Cloud speaks HTTPS rather than the native protocol, so its
+	// credentials show up as ordinary basic-auth URLs. The domain is what makes
+	// this unambiguous - a bare https:// URL with credentials is not a
+	// ClickHouse finding.
+	cloudPat = regexp.MustCompile(`\bhttps://([^\s:/@]{0,64}):([^\s:/@]{3,100})@([-\w.]+\.clickhouse\.cloud(?::\d{1,5})?)`)
```
[the remark this answers is not in the document]
Comment (DrFaust92) [reply pkg/detectors/clickhouse/clickhouse.go]: This one is addressed — fixed in 12b4396, I just replied in the PR conversation rather than on this thread, so it still reads as open.

go-re2 has no lookahead, so instead of anchoring in the pattern the host is now captured whole and the domain checked with a suffix test in code. Your exact example returns nothing now:

```
https://user:pass@foo.clickhouse.cloud.internal          -> 0 results
https://user:pass@foo.clickhouse.cloud.example.com:8443  -> 0 results
https://user:pass@real.clickhouse.cloud:8443             -> 1 result, host real.clickhouse.cloud:8443
```

Worth adding that I think this deserved more than Low. The consequence was not just a wrong host in the output — verification would have sent the credential to a ClickHouse Cloud hostname it does not belong to, so a secret for an internal host would have been transmitted to a third party.

Fixing it did introduce the trailing-dot regression you caught in r4066766815, since moving the check out of the pattern let the host group swallow a sentence-ending period. That is fixed in af17c66.

File: pkg/detectors/clickhouse/clickhouse.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	dataStr := string(data)
+
+	var results []detectors.Result
+	seen := make(map[string]struct{})
+
+	for _, m := range append(keyPat.FindAllStringSubmatch(dataStr, -1), cloudPat.FindAllStringSubmatch(dataStr, -1)...) {
+		user, password, hostPort := m[1], m[2], m[3]
+		if user == "" {
+			user = defaultUser
+		}
```
[the remark this answers is not in the document]
Comment (DrFaust92) [reply pkg/detectors/clickhouse/clickhouse.go:89]: Addressed in 12b4396 — replying here too, since I answered in the PR conversation and this thread still reads as open.

This was the most serious of the three and I confirmed it against a real server rather than from the code. With a user whose password is `p@ssw0rd!x`, ClickHouse returns `200` for the raw value and `403` for `p%40ssw0rd%21x`. Since the pattern's character class excludes a literal `@`, the escaped form is the only way such a password can appear in a connection string — so a correctly written, working credential was being reported as determinately invalid.

Userinfo is now percent-decoded before it is used for verification, `SecretParts` or `Raw`. `Redacted` still masks the password as written, and a value that is not validly encoded is left alone so a password containing a bare `%` still reports correctly.

File: pkg/detectors/clickhouse/clickhouse.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+		return false, err
+	}
+	defer func() {
+		_, _ = io.Copy(io.Discard, res.Body)
+		_ = res.Body.Close()
+	}()
+
+	switch res.StatusCode {
+	case http.StatusOK:
+		return true, nil
```
[the remark this answers is not in the document]
Comment (DrFaust92) [reply pkg/detectors/clickhouse/clickhouse.go:249]: Addressed in 12b4396 — replying here too, since I answered in the PR conversation and this thread still reads as open.

You were right that nothing tied the responder to ClickHouse. Verification now requires an `X-ClickHouse-*` header before reading anything into the status code. I checked that real replies carry them in both directions: a 200 returns `X-ClickHouse-Query-Id` and `X-ClickHouse-Summary`, and a 403 returns `X-ClickHouse-Exception-Code: 516`. Without one the result is indeterminate.

That also covers the converse case, which the finding did not mention but matters as much: a 403 from an unrelated server would previously have read as a rejected credential and marked a live secret invalid.
